qemu/tests/block_hotplug.py

```diff
@@ -21,144 +22,179 @@ def run(test, params, env):
-                                      logging.info)
-                if params.get("disk_op_cmd"):
-                    if params.get("os_type") == "linux":
-                        test_cmd = params.get("disk_op_cmd") % (disk, disk)
-                    elif params.get("os_type") == "windows":
-                        test_cmd = params.get("disk_op_cmd") % (disk_letter[num],
-                                                                disk_letter[num])
-                        test_cmd = utils_misc.set_winutils_letter(session, test_cmd)
-                    else:
-                        test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
-
-                    status, output = session.cmd_status_output(test_cmd,
-                                                               timeout=disk_op_timeout)
-                    if status:
-                        test.fail("Check for block device failed."
-                                  "Output: %s" % output)
-                session.close()
-
-                devs = [dev for dev in devs if not isinstance(dev, qdevices.QDrive)]
-                device_list.extend(devs)
+        devices = []
+        run_sub_test("before_plug")
+        for index in range(1, data_disks_num):
+            if params.get("need_plug", 'no') == "yes":
+                devs = create_block_devices(imgs[index])
+                disk = hotplug_block_devices(devs)
+                if disk_op_cmd:
+                    disk_io_test()
+                get_block_devices(devs)
             else:
-                for device in vm.devices:
-                    if device.get_param("id") == img_list[num + 1]:
-                        device_list.append(device)
-
-        plug_tag = "after_plug"
-        vm_switched_off = run_sub_test(params, plug_tag)
-        if vm_switched_off:
-            return
-
-        plug_tag = "before_unplug"
-        run_sub_test(params, plug_tag)
-
-        error_context.context("Unplug block device (iteration %d)" % iteration,
-                              logging.info)
-        disks_before_unplug = find_disk(vm, get_disk_cmd)
-        for device in reversed(device_list):
-            ret = vm.devices.simple_unplug(device, vm.monitor)
-            if ret[1] is False:
-                test.fail("Failed to unplug device '%s'."
-                          "Output:\n%s" % (device, ret[0]))
-
-        unplug_disks = utils_misc.wait_for(lambda: get_new_disk(find_disk(vm, get_disk_cmd),
-                                           disks_before_unplug), pause)
-        if len(unplug_disks) != blk_num:
-            test.fail("Failed to unplug devices from guest, need to unplug: %d,"
-                      "actual unplug: %d" % (blk_num, len(unplug_disks)))
-
-        plug_tag = "after_unplug"
-        run_sub_test(params, plug_tag)
+                get_block_devices(imgs[index])
+        run_sub_test("after_plug")
+        if not vm.is_alive():
+            break
```

Comment:
I think it is reasonable in original implementation to judge whether the vm is intended down here.